src/network/networkd-manager.c

```diff
@@ -150,6 +150,16 @@ int manager_connect_bus(Manager *m) {
                 return 0;
         }
 
+        /* Do we have a pending hostname from DHCP while D-Bus wasn't up yet? */
+        if (m->pending_hostname) {
+                assert(m->pending_hostname_link);
+                r = link_set_hostname(m->pending_hostname_link, m->pending_hostname);
+                if (r < 0)
+                        log_link_error_errno(m->pending_hostname_link, r, "Failed to set transient hostname to '%s': %m", m->pending_hostname);
+                m->pending_hostname_link = NULL;
+                m->pending_hostname = mfree(m->pending_hostname);
+        }
```

Comment:
so, I think link_set_hostname() should be moved out of the "Link" context, and into the "Manager" context. i.e. chnage link_set_hostname() to become manager_set_hostname()... It's a system-wide operation it executes after all, not really a per-link operation...

I'd also do away with the "pending" concept. Simply have manager_set_hostname() do two things: store the hostname parameter in some field Manager::configured_hostname, and try to sync it onto hostnamed. And then, each time we gain a bus connection sync it again. I think this is ultimately simpler and more robust, and effectively the same. After all normally we should only get a bus connection once, and hence it's entirely safe to just do the sync hostname thing each time we connect to the bus...